### DISCUSSION TITLE
Why validation when not optimizing hyperparameters ?

### DISCUSSION THREAD

**Discussion by issamsouassi** [neuralhydrology/neuralhydrology#186] [Q&A]
Hi all!
I hope you're doing well.

It's been 3 months using Neural Hydrology on modeling streamflow (Qobs) in Maghreb region using a dataset of 98 watersheds based on different rainfall satellite products; ERA5, Chirps, Persiann, GPM. And T°min_era5, T°max_era5 and pet_hargreaves.

I have a problem with the validation process. As much as I know, the validation in Lstm is applied to optimize the model's hyperparameters, but in Neural Hydrology they're already fixed in the config file.

- Can you please explain the purpose of the validation process, can we do the work without it?
- Given the small amount of basins (98), how to better split Train/"Validation"/Test ?

I also have a question about time periods :
 
- The data availability of the target variable (Qobs) is different from a basin to another, what period of time to chose when using several basins with Qobs_period_availability different ?
(Already tried the period [15.7years] common between 84 basins, but the run isn't showing good results)

Thank you in advance.

**Comment by darylsuen** [comment]
Hi, I'm very new to everything so please take that into account. From my limited understanding, the model is adjusting the weights of various operations that are going on, maybe inside the neural network component of the model and probably the predictive head during training with the training dataset. Once it has gone through the dataset (an epoch), it will evaluate the current version of itself against the validation data. It can use this to make some adjustments before starting another round of training. The validation dataset should help the model prevent overfitting. While the model is doing this iterative process of training and evaluating it should see the loss function decrease (a metric like RMSE, for example) but it can start learning the training data too well so that it can make very good predictions during the training but those weights are not good for predicting on new data. When this happens the model should see a decrease in performance on the validation data (a symptom of overfitting) and that should bring training to a stop. Then the model will evaluate its final version of itself on the test dataset. In standalone scripts for LSTM that you write yourself, you can add things like early-stopping to also help prevent overfitting. Another hyperparameter to adjust for overfitting is the dropout rate. I apologize if you already know all of this but to answer one of your questions, I think that omitting the validation dataset would lead to overfitting and poor performance. I'm not sure if NH will allow a run without a validation dataset as I have not tried that before. 

I think you might also be at the point where you would try finetuning. Your model has been generally trained on many basins and this helps the model 'learn' real-world processes and how they relate to discharge rates. However, I think at a certain point you would want to finetune the training to maximize performance for a specific basin or a group of basins with similar characteristics. Also, if a basin is prone to highly volatile discharge rates it would probably be difficult to build a high performing model for that basin unless you can figure out what the most important factors are that govern discharge rates for that basin and can find variable data to represent those factors. For example, in an arid region prone to flash flooding during episodic rain events, perhaps topography is very important and you could try creating variables from Strahler stream order developed from a very high resolution DEM raster that captures undulations in the terrain. You could create a variable like total length of runoff channels for each Strahler order and maybe the total count of channels for each Strahler order. Strahler stream order is fairly easy to create in a GIS software like ArcGIS and I imagine QGIS has a tool or plugin that can do it, too.

You could play around with adding lagged features to add emphasis  previous variables values, for example, lagged features for all your variables for a month (shifting your data - check out this in pytorch. You could just write a script for adding lagged features to your forcing dataset, too, but there is a lagged_features option in NH but I haven't learned how to use it yet. 

I'm afraid I can't contribute to the question about varying data time periods among basins in the dataset.

Good luck!
